LiS/man/man4/clone.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: clone.4.man,v 1.1.4.6 2008-04-29 08:33:18 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:18 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: clone.4.man,v $
.\" Revision 1.1.4.6  2008-04-29 08:33:18  brian
.\" - update headers for Affero release
.\"
.\" Revision 1.1.4.5  2006/12/28 10:42:07  brian
.\" - a bit better make check target
.\"
.\" Revision 1.1.4.4  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH CLONE 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B /dev/clone
\- the STREAMS clone driver
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/types.h>
.br
.B #include <sys/stat.h>
.br
.B #include <fcntl.h>
.HP 8
.IB s " = open(" /dev/xxx_clone ", " 0 );
.\"
.\"
.SH DESCRIPTION
.PP
This driver is used to assist LiS in implementing the
.I clone
open function.  It appears under its
own name as
.BR /dev/clone_drvr .
By convention, it is allocated the first major number of all the
.IR STREAMS (4)
drivers.

To implement clone opens, one creates a node in the
.B /dev
directory for a device whose major number is set to that of the clone driver,
and whose minor number is the major number of the driver to which the clone open
is to be directed.

The clone driver's open routine forwards the open call to the target driver,
passing a unique flag that informs the driver that a clone open is being
requested.  The target driver then allocated a minor device number to uniquely
associate with this instance of the open operation.  The clone driver
synthesizes a new major/minor
.I device id
to pass back to LiS.  LiS recognizes the change of major/minor from the original
open and takes steps to allocate control structures unique to this open.

The
.I clone open
operation is intended to make is easy to open one device from a pool of devices,
such as pseudo ttys or logical connections.  It saves application programs from
having to scan a list of device mnemonics issuing trial opens until one is found
that succeeds.

Note that the driver is named
.B /dev/clone_drvr
instead of the more traditional SVR4
.BR /dev/clone .
This is to avoid a conflict with another driver named
.B /dev/clone
on Linux systems.
.\"
.\"
.\".SH CAVEATS
.\".PP
.\"
.\"
.\".SH BUGS
.\".PP
.\"
.\"
.SH DEVICES
.PP
.TP
.B /dev/clone_drvr
The clone driver device.
.\"
.\"
.\".SH "SEE ALSO"
.\".PP
.\"
.\"
.\".SH VERSIONS
.\".PP
.\"
.\"
.XX
.[
magic
.]
.[
svr42
.]
.[
svr4
.]
.[
svr3
.]
.[
LiS
.]
.\"
.[
$LIST$
.]
.\"
.SH AUTHOR
.PP
David Grothe
.RB < dave@gcom.com >
.\"
.TI


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/drivers/str/clone.c
/*****************************************************************************

 @(#) $RCSfile: clone.c,v $ $Name:  $($Revision: 1.1.1.2.4.11 $) $Date: 2008-04-29 08:33:11 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:11 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: clone.c,v $
 Revision 1.1.1.2.4.11  2008-04-29 08:33:11  brian
 - update headers for Affero release

 Revision 1.1.1.2.4.10  2007/08/14 10:46:56  brian
 - GPLv3 header update

 *****************************************************************************/

#ident "@(#) $RCSfile: clone.c,v $ $Name:  $($Revision: 1.1.1.2.4.11 $) $Date: 2008-04-29 08:33:11 $"

static char const ident[] = "$RCSfile: clone.c,v $ $Name:  $($Revision: 1.1.1.2.4.11 $) $Date: 2008-04-29 08:33:11 $";

/************************************************************************
*                          Clone Driver                                 *
*************************************************************************
*									*
* Author:	David Grothe <dave@gcom.com>				*
*									*
* Copyright (C) 1997  David Grothe, Gcom, Inc <dave@gcom.com>		*
*									*
* This driver's open routine divides the device number passed in into	*
* the major and minor parts.  It treats the minor number as the major	*
* number of the driver that the user really wants to open.  It then	*
* calls that driver's open routine passing in the target driver's	*
* major and minor zero.  It sets the clone option bit in this open.	*
*									*
* The queue passed to clone is given the attributes of the target	*
* driver rather than of the clone driver.				*
*									*
* When this driver returns from its open routine the dev_t structure	*
* passed to it will no doubt be modified.  The caller, perhaps clear	*
* back in the I/O system, must take care to notice this and allocate	*
* a new inode for this stream.						*
*									*
************************************************************************/

#include <sys/stream.h>
#include <sys/osif.h>

static struct module_info clone_minfo = {
	0,				/* id */
	"clone",			/* name */
	0,				/* min packet size accepted */
	0,				/* max packet size accepted */
	0,				/* high water mark */
	0				/* low water mark */
};

static int _RP clone_open(queue_t *, dev_t *, int, int, cred_t *);
static int _RP clone_close(queue_t *, int, cred_t *);

/* qinit structures (rd and wr side) 
 */
static struct qinit clone_rinit = {
	NULL,				/* put */
	NULL,				/* service */
	clone_open,			/* open */
	clone_close,			/* close */
	NULL,				/* admin */
	&clone_minfo,			/* info */
	NULL				/* stat */
};

static struct qinit clone_winit = {
	NULL,				/* put */
	NULL,				/* service */
	NULL,				/* open */
	NULL,				/* close */
	NULL,				/* admin */
	&clone_minfo,			/* info */
	NULL				/* stat */
};

/* streamtab for the clone driver.
 */
struct streamtab clone_info = {
	&clone_rinit,			/* read queue */
	&clone_winit,			/* write queue */
	NULL,				/* mux read queue */
	NULL				/* mux write queue */
};

static int _RP
clone_open(queue_t *q, dev_t *devp, int flag, int sflag, cred_t *credp)
{
	int major_dev;
	int rslt;
	streamtab_t *st;

	if (sflag == CLONEOPEN)
		return (OPENFAIL);

	/* 
	 * Minor passed to us functions as major for the target driver
	 */
	major_dev = getminor(*devp);
	st = lis_find_strdev(major_dev);
	if (st == NULL)
		return (ENOENT);	/* no such driver */

	*devp = makedevice(major_dev, 0);	/* construct new major */

	if (st->st_rdinit->qi_qopen == NULL)
		return (OPENFAIL);
	lis_setq(q, st->st_rdinit, st->st_wrinit);	/* xfer queue params */
	rslt = st->st_rdinit->qi_qopen(q, devp, flag, CLONEOPEN, credp);
	if (rslt != 0)
		return (rslt);

	return (0);		/* success */

}				/* clone_open */

static int _RP
clone_close(queue_t *q, int dummy, cred_t *credp)
{
	(void) q;		/* compiler happiness */
	(void) dummy;
	(void) credp;

	return (0);
}

void
clone_init(void)
{
	/* 
	 *  indicate that this is a CLONE pseudo device
	 */
	LIS_DEVFLAGS(LIS_CLONE) |= LIS_MODFLG_CLONE;
}

#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-" __stringify(LIS_OBJNAME));
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0));
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0) "-*");
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0) "-0");
MODULE_ALIAS("/dev/clone_drvr");
#endif
